src/midi/midi_soundcanvas.cpp

```diff
@@ -0,0 +1,729 @@
+		}
+		had_underruns = true;
+	}
+
+	static std::vector<AudioFrame> audio_frames = {};
+
+	const auto has_dequeued = audio_frame_fifo.BulkDequeue(audio_frames,
+	                                                       requested_audio_frames);
+
+	if (has_dequeued) {
+		mixer_channel->AddSamples_sfloat(requested_audio_frames,
+		                                 &audio_frames[0][0]);
+
+		last_rendered_ms = PIC_FullIndex();
+	} else {
+		assert(!audio_frame_fifo.IsRunning());
+		mixer_channel->AddSilence();
+	}
+}
+
+void MidiDeviceSoundCanvas::RenderAudioFramesToFifo(const int num_audio_frames)
+{
+	assert(num_audio_frames > 0);
+
+	static std::vector<float> left  = {};
+	static std::vector<float> right = {};
+
+	// Maybe expand the vectors
+	if (check_cast<int>(left.size()) < num_audio_frames) {
+		left.resize(num_audio_frames);
+		right.resize(num_audio_frames);
+	}
+
+	float* audio_out[] = {left.data(), right.data()};
+
+	clap.plugin->Process(audio_out, num_audio_frames, clap.event_list);
+	clap.event_list.Clear();
+
+	for (auto i = 0; i < num_audio_frames; ++i) {
+		audio_frame_fifo.Enqueue({left[i], right[i]});
+	}
+}
+
+// The next MIDI work task is processed, which includes rendering audio frames
+// prior to sending channel and sysex messages to the plugin
+void MidiDeviceSoundCanvas::ProcessWorkFromFifo()
+{
+	const auto work = work_fifo.Dequeue();
+	if (!work) {
+		return;
+	}
+
+#if 0
+	// To log inter-cycle rendering
+	if (work->num_pending_audio_frames > 0) {
+		LOG_MSG("SOUNDCANVAS: %2u audio frames prior to %s message, followed by "
+		        "%2lu more messages. Have %4lu audio frames queued",
+		        work->num_pending_audio_frames,
+		        work->message_type == MessageType::Channel ? "channel" : "sysex",
+		        work_fifo.Size(),
+		        audio_frame_fifo.Size());
+	}
+#endif
+
+	if (work->num_pending_audio_frames > 0) {
+		RenderAudioFramesToFifo(work->num_pending_audio_frames);
+	}
+
+	if (work->message_type == MessageType::Channel) {
+		assert(work->message.size() >= MaxMidiMessageLen);
+		clap.event_list.AddMidiEvent(work->message, 0);
+
+	} else {
+		assert(work->message_type == MessageType::SysEx);
+		clap.event_list.AddMidiSysExEvent(work->message, 0);
+	}
+}
+
+// Keep the FIFO populated with freshly rendered buffers
+void MidiDeviceSoundCanvas::Render()
+{
+	while (work_fifo.IsRunning()) {
+		work_fifo.IsEmpty() ? RenderAudioFramesToFifo()
+		                    : ProcessWorkFromFifo();
+	}
+}
+
+void SOUNDCANVAS_ListDevices(MidiDeviceSoundCanvas* device, Program* caller)
+{
+	using namespace SoundCanvas;
+
+	// Table layout constants
+	constexpr auto ColumnDelim = " ";
+	constexpr auto Indent      = "  ";
+
+	const auto available_models = [&] {
+		std::set<const SynthModel*> available_models = {};
+
+		for (auto m : all_models) {
+			if (const auto p = try_load_plugin(*m); p.plugin) {
+				available_models.insert(m);
+			}
+		}
+		return available_models;
+	}();
```

Comment:
This is rather slow.  `mixer /listmidi` takes about 2 seconds to return on my machine and I benchmarked this part as taking 750ms of that time.  Don't you have this information already cached in `PluginInfo` from when you enumerate the plugins at the start?

Reply:
Yeah, I'll cache it, good point. It's reasonable to cache everything at first lookup, then if people change their plugins directory while Staging is running they'll need a restart for the changes to take effect.